timelock-server/src/main/java/com/palantir/atlasdb/timelock/clock/ClockSkewMonitor.java

```diff
@@ -0,0 +1,174 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.clock;
+
+import static java.lang.Thread.sleep;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import javax.net.ssl.SSLSocketFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.http.AtlasDbHttpClients;
+import com.palantir.atlasdb.http.ExceptionRetryBehaviour;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.logsafe.SafeArg;
+
+/**
+ * ClockSkewMonitor keeps track of the system time of the other nodes in the cluster, and compares it to the local
+ * clock. It's purpose is to monitor if the other nodes' clock progress at the same pace as the local clock.
+ */
+public final class ClockSkewMonitor implements Runnable {
+
+    private static final long sleepTimeMillis = 1_000; // 1 s
+    private static final long millisToNanos = 1_000_000; // 10^6
+    private static final long maxTimeSincePreviousRequestNanos = 10 * sleepTimeMillis * millisToNanos; // 10 s
+    private static final long maxRequestTimeNanos = 10_000_000; // 10 ms
+
+    private static final MetricRegistry metricRegistry = AtlasDbMetrics.getMetricRegistry();
+
+    private final Logger logger = LoggerFactory.getLogger(ClockSkewMonitor.class);
+    private final Map<String, ClockService> monitors;
+    private final Map<String, RequestTime> previousRequests;
+    private final Supplier<Boolean> shouldRunClockSkewMonitor;
+
+    public static ClockSkewMonitor create(Set<String> remoteServers, Optional<SSLSocketFactory> optionalSecurity,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        Map<String, ClockService> monitors = Maps.toMap(remoteServers,
+                (remoteServer) -> AtlasDbHttpClients.createProxy(optionalSecurity, remoteServer, ClockService.class));
+
+        Map<String, RequestTime> previousRequests = Maps.newHashMap();
+        for (String server : remoteServers) {
+            previousRequests.put(server, null);
+        }
+
+        return new ClockSkewMonitor(monitors, previousRequests, shouldRunClockSkewMonitor);
+    }
+
+    private ClockSkewMonitor(
+            Map<String, ClockService> monitors,
+            Map<String, RequestTime> previousRequests,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        this.monitors = monitors;
+        this.previousRequests = previousRequests;
+        this.shouldRunClockSkewMonitor = shouldRunClockSkewMonitor;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            try {
+                sleep(sleepTimeMillis);
+                if (shouldRunClockSkewMonitor.get()) {
+                    runInternal();
+                }
+            } catch (InterruptedException e) {
```

Comment:
we don't have to catch this if the executor service is registered with dropwizard; it will be shut down automatically

though we DO want to `catch(Throwable t)` and log a warning; otherwise an uncaught exception will cause the executor to stop running